### PR TITLE
Redis

### PR DESCRIPTION
**Pull request by gznznzjsn**

### CODE REVIEW

File: src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.springframework.data.redis.core.ReactiveHashOperations;
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.util.function.Tuple2;
+
+import java.util.List;
+
+@Service
+public class RequirementServiceCachedImpl extends RequirementServiceImpl {
```
[the remark this answers is not in the document]
Comment (h1alexbel) [reply src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java:16]: read https://stackoverflow.com/questions/49002/prefer-composition-over-inheritance

File: src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    public Mono<Requirement> create(Requirement requirement) {
+        return super.create(requirement)
+                .flatMap(r -> Mono.zip(
+                        Mono.just(r),
+                        hashOps.remove(KEY, r.getTask().getId())
+                ))
+                .map(Tuple2::getT1);
+    }
+
+    private Mono<List<Requirement>> getFromRepositoryAndCache(String taskId) {
```
Comment (h1alexbel) [inline src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java:49]: use `Flux` here

**Review by h1alexbel — CHANGES_REQUESTED** [review]

File: src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gznznzjsn) [reply src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java]: Done

File: src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java (hunk 2 above)
Comment (gznznzjsn) [reply src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java]: Done

File: src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java [hunk 3]
```diff
@@ -12,45 +12,44 @@
...
-    @Autowired
-    public RequirementServiceCachedImpl(
-            RequirementRepository requirementRepository,
-            RequirementSender requirementSender,
-            ReactiveHashOperations<String, String, List<Requirement>> hashOps
-    ) {
-        super(requirementRepository, requirementSender);
-        this.hashOps = hashOps;
-    }
+    private final RequirementServiceImpl requirementService;
```
Comment (h1alexbel) [inline src/main/java/com/gznznzjsn/taskservice/service/impl/RequirementServiceCachedImpl.java:23]: don't hardcode your dependencies using concrete classes

**Review by h1alexbel — CHANGES_REQUESTED** [review]

**Review by h1alexbel — APPROVED** [review]

**Comment by h1alexbel** [comment]
@gznznzjsn cheers!

**Pull request merged**